Sprite scale9

ricardoquesada opened the pull request:
Adds support for "sliced 9" sprite to `Sprite`.  This will deprecate `Scale9Sprite`.

It adds two new APIs to `Sprite`:
- `setCenterRect(Rect())`  // in pixel coordinates
- `setCenterRectNormalized(Rect())`   // in normalized coordinates (from 0 to 1)

To stretch the sprite, just call `sprite->setContentSize()`.
`setContentSize()` will stretch the sprite both in 9-slice mode, or "normal" mode.

ricardoquesada committed:
Adds slice9 support for Sprite.

how to use it:

// points coords
sprite->setCenterRect(Rect(x,y,w,h));

// normalized coords
sprite->setCenterRectNormalized(Rect(x,y,w,h));
starts scale9sprite in sprite

more slice 9 changes

sprite 9 slice works?

kind of works

correct anchor point

slice 9 works, at least with non-rotated atlases

streched works ok

better Y invert code.

cleaner, compatible with the previous code

yay, scaling workings...

need a better api now

sets scale correctly

yay! works as expected!

more fixes and tests

better test for box

setContentSize() changes size in non-9-slice mode sprites as well

setCenterRect() -> setCenterRectNormalized()

yet another test

adds setPositionNormalized()

adds setCenterRect() tests

remove devel team from xcode

tests: add one more tests

fun test!

improved test

yet another test for slice 9

dumganhar commented on cocos/2d/CCAutoPolygon.h:
```diff
@@ -81,6 +81,15 @@ class CC_DLL PolygonInfo
+     * as the verts memory are managed by other objects
+     * @param quad  a pointer to the V3F_C4B_T2F_Quad quads
+     */
+    void setQuads(V3F_C4B_T2F_Quad *quads, int nubmerOfQuads);
```
typo: nubmerOfQuads -> numberOfQuads

ricardoquesada replied:
fixed. thanks.

minggo commented on cocos/2d/CCAutoPolygon.cpp:
```diff
@@ -36,6 +36,18 @@ THE SOFTWARE.
+    0+4*6,1+4*6,2+4*6, 3+4*6,2+4*6,1+4*6,
+    0+4*7,1+4*7,2+4*7, 3+4*7,2+4*7,1+4*7,
+    0+4*8,1+4*8,2+4*8, 3+4*8,2+4*8,1+4*8,
+};
```
Does it have 36 vertexes? I think it only has 16 vertexes? Am i wrong?

ricardoquesada replied:
It has the indices for 9 quads, and each quad requires 6 indices and 4 vertices, so it needs: 9 \* 6 = 54 indices and 9*4=36 vertices.

dumganhar replied:
![scale9sprite](https://cloud.githubusercontent.com/assets/493372/19584070/465274ea-9774-11e6-9530-3b1c5ef5fc15.jpg)

I guess there are only 16 vertices and 4 x 9 = 36 indices (one quad requires 4 indices, there are 9 quads, so there are 4 x 9 = 36 indices). @ricardoquesada , anything I missed?

ricardoquesada replied:
you are right. I'm just reusing the Sprite's quad logic.

minggo commented on cocos/2d/CCSprite.cpp:
```diff
@@ -410,50 +416,114 @@ void Sprite::setTextureRect(const Rect& rect, bool rotated, const Size& untrimme
 
-    float relativeOffsetX = _unflippedOffsetPositionFromCenter.x;
-    float relativeOffsetY = _unflippedOffsetPositionFromCenter.y;
+    // overrides might update _rect. Use _rect instead of _rect starting from here
```
> Use _rect instead of _rect

What does it mean?

ricardoquesada replied:
deprecated comment. I'll remove it.

ricardoquesada commented on cocos/2d/CCSprite.cpp:
```diff
@@ -410,50 +416,114 @@ void Sprite::setTextureRect(const Rect& rect, bool rotated, const Size& untrimme
 
-    float relativeOffsetX = _unflippedOffsetPositionFromCenter.x;
-    float relativeOffsetY = _unflippedOffsetPositionFromCenter.y;
+    // overrides might update _rect. Use _rect instead of _rect starting from here
```
bug: it should read: "Use _rect instead of rect"

ricardoquesada committed:
fixes anchorPoint issues

ricardoquesada committed:
adds documentation

ricardoquesada commented:
@minggo @dumganhar 
are you Ok if I merge this PR? or are you still reviewing ?

The thing that you mentioned about the Triangles vs. Quad is valid. I'll try to implement after I finish the Creator for C++. 

thanks.

dumganhar commented:
@ricardoquesada , it looks good for me. 
@minggo , did you finish the reviewing?

ricardoquesada committed:
fix: using top-left coordinate for setRect

ricardoquesada committed:
sprite: fixes related to scale9 and tiled

ricardoquesada commented:
@dumganhar thanks. I've just added a few more fixes.

@minggo any update? thanks.

minggo commented:
@ricardoquesada sorry for late reply. Looks good to me.

ricardoquesada commented:
@minggo @dumganhar ok. thanks. I'll merge it tomorrow (Tuesday)

ricardoquesada committed:
Sprite: slice 9 fixes

works Ok with rotated frames
uses `setCapInsets` instead of `setCenterRect` to be more familiar
with `UIScale9Sprite`

ricardoquesada committed:
fixes js and lua bindings for Autopolygon
